File: sdk-mobile/ios/AgeKeySwift/Sources/AgeKeySwift/AgeKeyClient.swift
import Foundation

public final class AgeKeyClient: @unchecked Sendable {
    private let config: AgeKeyConfig
    private let session: URLSession
    private let encoder: JSONEncoder
    private let decoder: JSONDecoder

    public init(config: AgeKeyConfig, session: URLSession = .shared) {
        self.config = config
        self.session = session
        self.encoder = JSONEncoder()
        self.decoder = JSONDecoder()
    }

    public func createSession(
        _ request: AgeKeyCreateSessionRequest
    ) async throws -> AgeKeySession {
        try await post(
            path: "/functions/v1/verifications-session-create",
            body: request,
            response: AgeKeySession.self
        )
    }

    public func completeFallback(
        sessionId: String,
        declaration: AgeKeyCompleteFallbackDeclaration,
        signals: AgeKeyCompleteFallbackSignals = AgeKeyCompleteFallbackSignals()
    ) async throws -> AgeKeyCompleteResponse {
        let body = AgeKeyCompleteFallbackRequest(
            declaration: declaration,
            signals: signals
        )
        return try await post(
            path: "/functions/v1/verifications-session-complete/\(sessionId)",
            body: body,
            response: AgeKeyCompleteResponse.self
        )
    }

    public func verifyToken(
        token: String,
        expectedAudience: String? = nil
    ) async throws -> AgeKeyTokenVerifyResponse {
        try await post(
            path: "/functions/v1/verifications-token-verify",
            body: AgeKeyTokenVerifyRequest(
                token: token,
                expectedAudience: expectedAudience
            ),
            response: AgeKeyTokenVerifyResponse.self
        )
    }

    private func post<RequestBody: Encodable, ResponseBody: Decodable>(
        path: String,
        body: RequestBody,
        response: ResponseBody.Type
    ) async throws -> ResponseBody {
        guard let url = URL(string: path, relativeTo: config.environment.baseURL) else {
            throw AgeKeyError.invalidURL
        }

        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue(config.apiKey, forHTTPHeaderField: "X-AgeKey-API-Key")
        req.setValue(config.defaultLocale, forHTTPHeaderField: "Accept-Language")

        do {
            req.httpBody = try encoder.encode(body)
        } catch {
            throw AgeKeyError.encoding(error.localizedDescription)
        }

        let (data, urlResponse) = try await session.data(for: req)
        guard let http = urlResponse as? HTTPURLResponse else {
            throw AgeKeyError.invalidResponse
        }

        guard (200...299).contains(http.statusCode) else {
            let text = String(data: data, encoding: .utf8) ?? ""
            throw AgeKeyError.httpStatus(http.statusCode, text)
        }

        do {
            return try decoder.decode(ResponseBody.self, from: data)
        } catch {
            throw AgeKeyError.decoding(error.localizedDescription)
        }
    }
}

File: sdk-mobile/android/agekey-android/agekey/src/main/java/com/ecadigital/agekey/AgeKeyClient.kt
package com.ecadigital.agekey

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

class AgeKeyClient(private val config: AgeKeyConfig) {
    suspend fun createSession(request: AgeKeyCreateSessionRequest): AgeKeySession =
        withContext(Dispatchers.IO) {
            val json = post(
                path = "/functions/v1/verifications-session-create",
                body = request.toJson()
            )
            parseSession(json)
        }

    suspend fun completeFallback(
        sessionId: String,
        declaration: AgeKeyFallbackDeclaration,
        signals: AgeKeyFallbackSignals = AgeKeyFallbackSignals()
    ): AgeKeyCompleteResponse = withContext(Dispatchers.IO) {
        val body = JSONObject()
            .put("method", "fallback")
            .put(
                "declaration",
                JSONObject()
                    .put("age_at_least", declaration.ageAtLeast)
                    .put("consent", declaration.consent)
            )
            .put(
                "signals",
                JSONObject()
                    .put("captcha_token", signals.captchaToken)
                    .put("device_fingerprint", signals.deviceFingerprint)
            )

        val json = post(
            path = "/functions/v1/verifications-session-complete/$sessionId",
            body = body
        )
        parseComplete(json)
    }

    suspend fun verifyToken(token: String, expectedAudience: String? = null): AgeKeyTokenVerifyResponse =
        withContext(Dispatchers.IO) {
            val body = JSONObject()
                .put("token", token)
                .put("expected_audience", expectedAudience)

            val json = post(
                path = "/functions/v1/verifications-token-verify",
                body = body
            )
            parseVerify(json)
        }

    private fun post(path: String, body: JSONObject): JSONObject {
        val url = URL(config.environment.baseUrl.trimEnd('/') + path)
        val conn = (url.openConnection() as HttpURLConnection)
        conn.requestMethod = "POST"
        conn.setRequestProperty("Content-Type", "application/json")
        conn.setRequestProperty("Accept-Language", config.defaultLocale)
        conn.setRequestProperty("X-AgeKey-API-Key", config.apiKey)
        conn.doOutput = true

        conn.outputStream.use { out ->
            out.write(body.toString().toByteArray(Charsets.UTF_8))
        }

        val status = conn.responseCode
        val stream = if (status in 200..299) conn.inputStream else conn.errorStream
        val text = stream?.bufferedReader(Charsets.UTF_8)?.use { it.readText() } ?: ""

        if (status !in 200..299) {
            throw AgeKeyException("AgeKey HTTP $status: $text")
        }

        return JSONObject(text)
    }
}
